ClangSharpPInvokeGenerator/Program.cs

```diff
@@ -100,20 +105,36 @@ public static int Run(InvocationContext context)
                 foreach (var file in files)
                 {
                     var translationUnitError = CXTranslationUnit.Parse(createIndex, file, arr, Array.Empty<CXUnsavedFile>(), translationFlags, out CXTranslationUnit translationUnit);
+                    bool skipProcessing = false;
 
                     if (translationUnitError != CXErrorCode.CXError_Success)
                     {
-                        Console.WriteLine($"Error: '{translationUnitError}' for '{file}'.");
-                        var numDiagnostics = translationUnit.NumDiagnostics;
+                        Console.WriteLine($"Error: Parsing failed for '{file}' due to '{translationUnitError}'.");
+                        skipProcessing = true;
+                    }
+                    else if (translationUnit.NumDiagnostics != 0)
+                    {
+                        Console.WriteLine($"Diagnostics for '{file}':");
 
-                        for (uint i = 0; i < numDiagnostics; ++i)
+                        for (uint i = 0; i < translationUnit.NumDiagnostics; ++i)
                         {
-                            var diagnostic = translationUnit.GetDiagnostic(i);
-                            Console.WriteLine(diagnostic.Spelling.ToString());
-                            diagnostic.Dispose();
+                            using (var diagnostic = translationUnit.GetDiagnostic(i))
+                            {
+                                Console.Write("    ");
+                                Console.WriteLine(diagnostic.Format(CXDiagnosticDisplayOptions.CXDiagnostic_DisplayOption).ToString());
+
+                                skipProcessing |= (diagnostic.Severity == CXDiagnosticSeverity.CXDiagnostic_Error);
```

Comment:
Skipping processing on error/fatal diagnostics seems like the right default. Otherwise, you can end up with weird codegen due to missing types or clang misinterpreting a macro expansion as something else.